Sources/MongoSwift/BSON/Codec.swift

```diff
@@ -333,3 +333,580 @@ private func getBsonValue(_ value: Any) -> BsonValue? {
         preconditionFailure("Value \(value) with type \(type(of: value)) didn't match any expected types")
     }
 }
+
+/// `BsonDecoder` facilitates the decoding of BSON into semantic `Decodable` types.
+public class BsonDecoder {
+
+    /// Contextual user-provided information for use during decoding.
+    open var userInfo: [CodingUserInfoKey: Any] = [:]
+
+    /// Options set on the top-level decoder to pass down the decoding hierarchy.
+    fileprivate struct _Options {
+        let userInfo: [CodingUserInfoKey: Any]
+    }
+
+    /// The options set on the top-level decoder.
+    fileprivate var options: _Options {
+        return _Options(userInfo: userInfo)
+    }
+
+    /// Initializes `self`.
+    public init() {}
+
+    /// Decodes a top-level value of the given type from the given BSON representation.
+    ///
+    /// - parameter type: The type of the value to decode.
+    /// - parameter data: The BSON document to decode from.
+    /// - returns: A value of the requested type.
+    /// - throws: An error if any value throws an error during decoding.
+    public func decode<T: Decodable>(_ type: T.Type, from data: Document) throws -> T {
```

Comment:
was debating whether we should add a`decode` method that directly takes in `Data`, or just let people make the document themselves with `Document.init(fromBSON:)`. what do you think?

Reply:
👍